risc0/zkp/verify/taps.h

```diff
@@ -0,0 +1,259 @@
+
+private:
+  const impl::ComboData* data_;
+  size_t id_;
+};
+
+class ComboIterator {
+public:
+  ComboIterator(const impl::ComboData* data, size_t id) : data_(data), id_(id) {}
+  ComboRef operator*() const { return ComboRef(data_, id_); }
+  ComboIterator& operator++() {
+    id_++;
+    return *this;
+  }
+
+private:
+  const impl::ComboData* data_;
+  size_t id_;
+};
+
+namespace impl {
+
+struct TapSetData {
+  const impl::TapData* taps;
+  size_t groupBegin[kNumRegisterGroups + 1];
+  ComboData combos;
+};
+
+} // namespace impl
+
+template <typename It> class IteratorRange {
+public:
+  IteratorRange(It begin, It end) : begin_(begin), end_(end) {}
+  It begin() const { return begin_; }
+  It end() const { return end_; }
+
+private:
+  It begin_;
+  It end_;
+};
+
+class TapSetRef {
+private:
+  const impl::TapSetData* data_;
+
+public:
+  TapSetRef(const impl::TapSetData* data) : data_(data) {}
+  size_t tapsSize() { return data_->groupBegin[kNumRegisterGroups]; }
+  IteratorRange<TapIterator> taps() {
+    return IteratorRange<TapIterator>(data_->taps,
+                                      data_->taps + data_->groupBegin[kNumRegisterGroups]);
+  }
+  IteratorRange<RegIterator> regs() {
+    return IteratorRange<RegIterator>(data_->taps,
+                                      data_->taps + data_->groupBegin[kNumRegisterGroups]);
+  }
+  IteratorRange<TapIterator> groupTaps(RegisterGroup group) {
+    size_t groupID = static_cast<size_t>(group);
+    return IteratorRange<TapIterator>(data_->taps + data_->groupBegin[groupID],
+                                      data_->taps + data_->groupBegin[groupID + 1]);
+  }
+  IteratorRange<RegIterator> groupRegs(RegisterGroup group) {
+    size_t groupID = static_cast<size_t>(group);
+    return IteratorRange<RegIterator>(data_->taps + data_->groupBegin[groupID],
+                                      data_->taps + data_->groupBegin[groupID + 1]);
+  }
+  size_t groupSize(RegisterGroup group) {
+    size_t groupID = static_cast<size_t>(group);
+    size_t last = (data_->taps + data_->groupBegin[groupID + 1] - 1)->offset;
+    return last + 1;
+  }
+  size_t combosSize() const { return data_->combos.count; }
+  IteratorRange<ComboIterator> combos() {
+    return IteratorRange<ComboIterator>(ComboIterator(&data_->combos, 0),
+                                        ComboIterator(&data_->combos, data_->combos.count));
+  }
+  ComboRef getCombo(size_t id) { return ComboRef(&data_->combos, id); }
+};
+
+class TapSet {
+private:
+  void finalize();
+
+  // Stuff used only while building
+  bool finalized_;
+  using reg_t = std::set<size_t>;
+  using group_t = std::map<size_t, reg_t>;
+  using all_t = std::map<RegisterGroup, group_t>;
+  all_t all_;
+
+  // Stuff that exists post finalization
+  std::vector<impl::TapData> taps_;
+  std::vector<uint16_t> comboTaps_;
+  std::vector<uint16_t> comboBegin_;
+  impl::TapSetData data_;
```

Comment:
nit: Would prefer to drop trailing `_`, but I could be convinced otherwise. I just don't think we've been consistent about this so far.